### PR TITLE
Header de-/serialization, regtest and equihash validation

### PR DESCRIPTION
**Pull request by wilsonmeier**
- Implemented correct header de-/serialization for BTG. Syncing works for pre- and postfork blocks with variable header size.
- Added flag to gzip compress header files (not ready yet).
- Replced server list with localhost for now.
- Replaced block explorer list.
- Some reformating.
- Removed checkpoints. Checkpoints actually don't work as BTG validates by using the last 30 block headers from height to check and they are not included. Needs to be fixed.
- Added node regtest compatibility
- Reworked handling of chunks. This decouples chunks from difficulty adjustment for prefork headers and allows to request smaller chunks.
- Added equihash class, equihash params and unit tests
- Added equihash solution validation to header verification

### CODE REVIEW

File: lib/blockchain.py [hunk 1]
```diff
@@ -247,83 +394,175 @@ def read_header(self, height):
...
+        if size <= 3:
+            word >>= 8 * (3 - size)
+            ret = word
+        else:
+            ret = word
+            ret <<= 8 * (size - 3)
+
+        return ret
 
     def target_to_bits(self, target):
```
Comment (h4x3rotab) [inline lib/blockchain.py]: Maybe it's better to handle it in the standard way. This is my Python implementation of the original C++ code:

```python
def uint256_to_compact(u):
    assert u >= 0
    nsize = (bits(u) + 7) // 8
    if nsize <= 3:
        c = u << (8 * (3 - nsize))
    else:
        c = u >> (8 * (nsize - 3))
    if c & 0x00800000:
        c >>= 8
        nsize += 1
    assert (c & ~0x007fffff) == 0
    assert nsize < 256
    c |= nsize << 24
    return c
```

File: lib/blockchain.py [hunk 2]
```diff
@@ -146,71 +201,152 @@ def size(self):
...
-            raw_header = data[i*80:(i+1) * 80]
-            header = deserialize_header(raw_header, index*2016 + i)
             self.verify_header(header, prev_hash, target)
-            prev_hash = hash_header(header)
+
+            headers[height] = header
+            prev_hash = hash_header(header, height)
+            offset += header_size
+            height += 1
+            ser_chunks += serialize_header(header)
```
Comment (h4x3rotab) [inline lib/blockchain.py]: What's the difference between the original chunk and the re-serialized chunk?

File: lib/util.py [hunk 3]
```diff
@@ -486,7 +465,7 @@ def block_explorer_URL(config, kind, item):
     if not kind_str:
         return
     url_parts = [be_tuple[0], kind_str, item]
-    return ''.join(url_parts)
+    return "/".join(url_parts)
```
Comment (h4x3rotab) [inline lib/util.py]: These parts already have "/" suffix. Do you really want to join with another "/"?

File: lib/blockchain.py [hunk 4]
```diff
@@ -99,20 +145,23 @@ class Blockchain(util.PrintError):
...
+        self.catch_up = None
         self.checkpoint = checkpoint
-        self.checkpoints = bitcoin.NetworkConstants.CHECKPOINTS
+        self.checkpoints = NetworkConstants.CHECKPOINTS
         self.parent_id = parent_id
         self.lock = threading.Lock()
-        with self.lock:
-            self.update_size()
+        self._size = 0
+        self.update_size()
```
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/blockchain.py]: No as the lock is done in the encapsulated read_file function.
Before the changes self.update_size() didn't set a lock by itself, hence it was set by the calling functions.

File: lib/blockchain.py [hunk 5]
```diff
@@ -220,23 +356,34 @@ def swap_with_parent(self):
 
     def write(self, data, offset, truncate=True):
         filename = self.path()
-        with self.lock:
-            with open(filename, 'rb+') as f:
-                if truncate and offset != self._size*80:
-                    f.seek(offset)
-                    f.truncate()
+        current_offset, header_size = self.get_offset(self.size())
```
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/blockchain.py]: Lock is done in write_file

File: lib/bitcoin.py [hunk 6]
```diff
@@ -70,37 +71,142 @@ def read_json(filename, default):
...
         cls.SEGWIT_HRP = "tbtg"
-        cls.HEADERS_URL = "https://headers.electrum.org/testnet_headers"
+        cls.HEADERS_URL = "https://headers.bitcoingold.org/testnet_headers"
         cls.GENESIS = "000000000933ea01ad0ee984209779baaec3ced90fa3f408719526f8d77f4943"
-        cls.DEFAULT_PORTS = {'t':'51001', 's':'51002'}
+        cls.DEFAULT_PORTS = {'t': '51001', 's': '51002'}
         cls.DEFAULT_SERVERS = read_json('servers_testnet.json', {})
         cls.CHECKPOINTS = read_json('checkpoints_testnet.json', [])
+        cls.FORK_HEIGHT = 3000
+        cls.PREMINE_SIZE = 3000
```
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/bitcoin.py]: Will change it

File: lib/bitcoin.py [hunk 7]
```diff
@@ -70,37 +71,142 @@ def read_json(filename, default):
...
+        cls.POW_TARGET_TIMESPAN_LEGACY = 14 * 24 * 60 * 60
+
+    @classmethod
+    def set_regtest(cls):
+        cls.TESTNET = False
+        cls.REGTEST = True
+        cls.WIF_PREFIX = 0xef
+        cls.ADDRTYPE_P2PKH = 111
+        cls.ADDRTYPE_P2SH = 196
+        cls.SEGWIT_HRP = "btg"
```
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/bitcoin.py]: Will change it

File: lib/bitcoin.py [hunk 8]
```diff
@@ -70,37 +71,142 @@ def read_json(filename, default):
...
         cls.ADDRTYPE_P2SH = 196
         cls.SEGWIT_HRP = "tbtg"
-        cls.HEADERS_URL = "https://headers.electrum.org/testnet_headers"
+        cls.HEADERS_URL = "https://headers.bitcoingold.org/testnet_headers"
         cls.GENESIS = "000000000933ea01ad0ee984209779baaec3ced90fa3f408719526f8d77f4943"
-        cls.DEFAULT_PORTS = {'t':'51001', 's':'51002'}
+        cls.DEFAULT_PORTS = {'t': '51001', 's': '51002'}
         cls.DEFAULT_SERVERS = read_json('servers_testnet.json', {})
         cls.CHECKPOINTS = read_json('checkpoints_testnet.json', [])
+        cls.FORK_HEIGHT = 3000
```
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/bitcoin.py]: Will change it

File: lib/util.py (hunk 3 above)
Comment (wilsonmeier) [reply lib/util.py]: :/ how did this made its way into the code? Anyway, i'll remove it :D

File: lib/blockchain.py (hunk 2 above)
Comment (wilsonmeier) [reply lib/blockchain.py]: Can be optimized away. Will change it.

File: lib/blockchain.py [hunk 9]
```diff
@@ -247,83 +394,175 @@ def read_header(self, height):
...
+                i = 0
+
+                while i < NetworkConstants.POW_AVERAGING_WINDOW and first is not None:
+                    total += self.bits_to_target(int(first.get('bits'), 16))
+                    prev_height = height - i - 1
+                    first = headers[prev_height] if prev_height in headers else self.read_header(prev_height)
+                    i += 1
+
+                if first is None:
+                    new_target = pow_limit
```
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/blockchain.py]: Good point. WIll change it.

File: lib/blockchain.py (hunk 1 above)
Comment (wilsonmeier) [reply lib/blockchain.py]: Sure, no problem.

File: lib/bitcoin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/bitcoin.py]: Done

File: lib/bitcoin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/bitcoin.py]: Done

File: lib/bitcoin.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/bitcoin.py]: Done

File: lib/blockchain.py (hunk 1 above)
Comment (wilsonmeier) [reply lib/blockchain.py]: Done

File: lib/blockchain.py (hunk 2 above)
Comment (wilsonmeier) [reply lib/blockchain.py]: Done

File: lib/util.py (hunk 3 above)
Comment (wilsonmeier) [reply lib/util.py]: Done

File: lib/blockchain.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/blockchain.py]: Done

File: lib/bitcoin.py [hunk 10]
```diff
@@ -70,37 +71,142 @@ def read_json(filename, default):
...
+def difficulty_adjustment_interval():
+    return NetworkConstants.POW_TARGET_TIMESPAN_LEGACY // NetworkConstants.POW_TARGET_SPACING
+
+
+def get_header_size(height):
+    return NetworkConstants.HEADER_SIZE if height >= NetworkConstants.FORK_HEIGHT \
+        else NetworkConstants.HEADER_SIZE_LEGACY
+
+
+def get_powlimit(height):
```
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/bitcoin.py]: Sorry, obsolete function. Not used anymore. Will remove it.

[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/bitcoin.py]: Done

**Comment by wilsonmeier** [comment]
It's just a placeholder.

File: lib/blockchain.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (h4x3rotab) [reply lib/blockchain.py]: ack

File: lib/blockchain.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (h4x3rotab) [reply lib/blockchain.py]: ack

File: lib/util.py (hunk 3 above)
Comment (h4x3rotab) [reply lib/util.py]: Please test this before the release

Comment (wilsonmeier) [reply lib/util.py]: This has been reverted here https://github.com/BTCGPU/electrum/pull/6/commits/885ab6bb4954396f26d565e27ad50918dd8fb9c1

File: lib/blockchain.py [hunk 11]
```diff
@@ -146,71 +201,152 @@ def size(self):
...
+            raise BaseException("bits mismatch: %s vs %s" % (bits, int(header.get('bits'), 16)))
         if int('0x' + _hash, 16) > target:
             raise BaseException("insufficient proof of work: %s vs target %s" % (int('0x' + _hash, 16), target))
+        if is_postfork(block_height):
+            header_bytes = bytes.fromhex(serialize_header(header, False))
+            nonce = uint256_from_bytes(bfh(header.get('nonce'))[::-1])
+            solution = bfh(header.get('solution'))[::-1]
+            offset, length = var_int_read(solution, 0)
+            solution = solution[offset:]
+
```
[the remark this answers is not in the document]
Comment (wilsonmeier) [reply lib/blockchain.py]: Will add height check with one of the upcoming commits.

**Pull request merged**